### PR TITLE
substrate client

### PR DESCRIPTION
**Pull request by rawdaGastan**

### CODE REVIEW

File: substrate/identity.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        identity_type=key_pair.crypto_type,
+        address=key_pair.ss58_address,
+        public_key=key_pair.public_key,
+        uri=key_pair.derive_path,
+    )
+
+    return identity
+
+
+def new_identity_from_ed25519_key(private_key: bytes):
```
Comment (xmonader) [inline substrate/identity.py:41]: let's name it `from_ed25519_key` and make it a staticmethod

File: substrate/identity.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    """creates a new identity from an ed25519 private key
+
+    Args:
+        private_key (bytes): ed25519 private key
+
+    Returns:
+        Identity: user identity
+    """
+    crypto_type = KeypairType.ED25519
+    key_pair = Keypair.create_from_private_key(private_key, ss58_format=NETWORK, crypto_type=crypto_type)
```
Comment (xmonader) [inline substrate/identity.py:51]: it's very weird to use `NETWORK` as ss58 format, i think we can use a better name?

File: substrate/identity.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    return generate_identity_from_key_pair(key_pair)
+
+
+def new_identity_from_sr25519_phrase(phrase: str):
```
Comment (xmonader) [inline substrate/identity.py:55]: let's make it a staticmethod `from_sr25519_phrase` are phrases the same as mnemonics?

File: substrate/identity.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        phrase (str): given phrase to generate a key pair
+
+    Returns:
+        Identity: user identity
+    """
+    key_pair = Keypair.create_from_uri(phrase)
+    return generate_identity_from_key_pair(key_pair)
+
+
+def new_identity_from_sr25519_mnemonics(mnemonics: str):
```
Comment (xmonader) [inline substrate/identity.py:68]: let's make it a staticmethod `from_sr25519_mnemonics`, still not sure about the difference between phrase and mnemnoics

File: substrate/twin.py [hunk 5]
```diff
@@ -0,0 +1,150 @@
+"""Twin module"""
+import hashlib
+from substrateinterface import SubstrateInterface
+from .identity import Identity
+import ipaddress
+
+
+class Twin_Info:
```
Comment (xmonader) [inline substrate/twin.py:8]: rename to `TwinInfo`, using underscores in classnames is highly discouraged

File: substrate/twin.py [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+    """Twin info class"""
+
+    def __init__(self, version: int, id: int, account_id: bytes, ip: ipaddress.IPv4Address, entities: list):
```
Comment (xmonader) [inline substrate/twin.py:11]: let's convert to a dataclass

File: substrate/twin.py [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+
+class Twin:
+    """Twin class"""
+
+    def __init__(self, substrate: SubstrateInterface, identity: Identity):
+        self.substrate = substrate
+        self.identity = identity
+        self.twin_info = None
+
+    def get(self):
```
Comment (xmonader) [inline substrate/twin.py:27]: I suggest you make staticmethods `from_publickey` and `from_id` methods

File: substrate/twin.py [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+        """get the twin info for the account ID
+
+        Returns:
+            Twin_Info: the info for a twin
+        """
+        twin_id = get_twin_by_public_key(self.identity.public_key, self.substrate)
+        if twin_id != None:
+            return get_twin_by_id(twin_id, self.substrate)
+
```
Comment (xmonader) [inline substrate/twin.py:36]: what happens if it's None?

File: substrate/twin.py [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+    def create(self, ip: str):
```
Comment (xmonader) [inline substrate/twin.py:37]: for consistency, change the ip in the constructor to be a string and do validations on that string, IPs should be str for ease of use

File: substrate/twin.py [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+        Raises:
+            Exception: creating a new twin failed
+        """
+        call = self.substrate.compose_call("TfgridModule", "create_twin", {"ip": ip})
+
+        extrinsic = self.substrate.create_signed_extrinsic(call, self.identity.key_pair)
+        result = self.substrate.submit_extrinsic(extrinsic, True, True)
+
+        if not result.is_success or result.error_message != None:
+            raise Exception("creating a new twin failed with error: ", result.error_message)
```
Comment (xmonader) [inline substrate/twin.py:52]: Create a custom exception TwinCreationException to be raised here instead

File: substrate/twin.py [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+        Raises:
+            Exception: updating a new twin failed
+        """
+        call = self.substrate.compose_call("TfgridModule", "update_twin", {"ip": ip})
+
+        extrinsic = self.substrate.create_signed_extrinsic(call, self.identity.key_pair)
+        result = self.substrate.submit_extrinsic(extrinsic, True, True)
+
+        if not result.is_success or result.error_message != None:
+            raise Exception("updating a new twin failed with error: ", result.error_message)
```
Comment (xmonader) [inline substrate/twin.py:69]: Let's have a custom error `TwinUpdateException`

File: substrate/twin.py [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+            "TfgridModule",
+            "user_accept_tc",
+            {"document_link": document_link, "document_hash": document_hash},
+        )
+
+        extrinsic = self.substrate.create_signed_extrinsic(call, self.identity.key_pair)
+        result = self.substrate.submit_extrinsic(extrinsic, True, True)
+
+        if not result.is_success or result.error_message != None:
+            raise Exception("accepting terms and conditions failed with error: ", result.error_message)
```
Comment (xmonader) [inline substrate/twin.py:97]: let's have a custom exception for it as well

File: substrate/twin.py [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+
+
+def get_twin_by_id(id: int, substrate: SubstrateInterface):
```
Comment (xmonader) [inline substrate/twin.py:100]: flip the arguments to have substrate as the first parameter

File: substrate/twin.py [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+
+    Raises:
+        Exception: twin not found
+
+    Returns:
+        Twin_Info: the info for a twin
+    """
+    twin = substrate.query("TfgridModule", "Twins", [id])
+    if twin["id"].value == 0:
+        raise Exception("twin not found")
```
Comment (xmonader) [inline substrate/twin.py:115]: Should be `ValueError` and need to encode the data in the exception as well

File: substrate/twin.py [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+    version = twin["version"].value
+    twin_id = twin["id"].value
+    account_id = twin["account_id"].value
+    ip = twin["ip"].value
+    entities = twin["entities"].value
+
+    return Twin_Info(version, twin_id, account_id, ip, entities)
+
+
+def get_twin_by_public_key(public_key: bytes, substrate: SubstrateInterface):
```
Comment (xmonader) [inline substrate/twin.py:126]: flip the parameters order

File: substrate/twin.py [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+    """get twin ID from a public key
+
+    Args:
+        public_key (bytes): identity public key
+        substrate (SubstrateInterface): substrate client
+    """
+
+    twin_id = substrate.query("TfgridModule", "TwinIdByAccountID", [public_key])
+    if twin_id == 0:
+        raise Exception("twin not found")
```
Comment (xmonader) [inline substrate/twin.py:136]: `ValueError` or a custom `TwinNotFoundException`

File: substrate/twin.py [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+
+    return twin_id
+
+
+def signed_terms_and_conditions(public_key: bytes, substrate: SubstrateInterface):
```
Comment (xmonader) [inline substrate/twin.py:141]: flip the parameters

File: test/substrate/account_test.py [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+
+def test_get_account_by_identity():
+    """test get account with an identity"""
+
+    substrate = start_local_connection()
+
+    test_identity = identity.new_identity_from_sr25519_mnemonics(ALICE_MNEMONICS)
+    test_account = account.Account(substrate, test_identity)
+
+    try:
```
Comment (xmonader) [inline test/substrate/account_test.py:18]: this test, is either passes or we need to assert it raises an exception

**Review by xmonader — CHANGES_REQUESTED** [review]
Added a few comments, let's address before going further

File: modules/vm.py [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+    ygg_ip: str
+    ip: str
+    description: str
+    cpu: int
+    memory: int
+    rootfs_size: int
+    entry_point: str
+    mounts: list
+    zlogs: list
+    env_vars: map
```
Comment (xmonader) [inline modules/vm.py:30]: what is the type map? did you mean dict?

Comment (rawdaGastan) [reply modules/vm.py:30]: please ignore this file as well as anything outside substrate

**Pull request merged**